openstack_dashboard/cc_websso_views.py

```diff
@@ -0,0 +1,98 @@
+import logging
+from django.contrib import auth
+from django.contrib.auth import views as django_auth_views
+from django.conf import settings
+from django import http as django_http
+from django.views.decorators.cache import never_cache
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.csrf import csrf_protect
+from django.views.decorators.debug import sensitive_post_parameters
+from openstack_auth import utils
+from openstack_auth import exceptions
+from openstack_auth import user as auth_user
+import six
+
+LOG = logging.getLogger(__name__)
+
+"""
+    The below overrides the defaul login functionality and redirects to auth with the preconfigured portal with
+    all the required parameters for login to this Horizon instance
+"""
+@sensitive_post_parameters()
+@csrf_protect
+@never_cache
+def login(request, template_name=None, extra_context=None, **kwargs):
+    if not request.user.is_authenticated():
+        host = getattr(settings, 'SSO_CALLBACK_HOST', None)
+        cc_portal_url = getattr(settings, 'CHAMELEON_PORTAL_SSO_BASE_URL', None) + getattr(settings, 'CHAMELEON_PORTAL_SSO_LOGIN_PATH', None)
+        if(host is None or cc_portal_url is None):
+            LOG.error('Misconfigured CC Portal SSO, settings:, '
+                + 'CHAMELEON_PORTAL_SSO_URL: ' + str(cc_portal_url) + ', SSO_CALLBACK_HOST: ' + str(host))
+            raise Exception('SSO Login Error')
+        login_url = cc_portal_url + '?host=' + host
+        if request.GET.get('next'):
+            login_url += '&next=' + request.GET.get('next')
+        if getattr(settings, 'WEBROOT', None) and getattr(settings, 'WEBROOT', None) != '/':
+            login_url += '&webroot=' + getattr(settings, 'WEBROOT', '')
+        return django_http.HttpResponseRedirect(login_url)
+    return openstack_auth.views.login(request, template_name=None, extra_context=None, **kwargs)
+
+"""
+    this is a custom websso endpoint for chameleon portal
+    This primary difference between this and the original websso login code is
+    that the original request's HTTP_REFERER is not used, in the standard websso impmlementation,
+    the HTTP_REFERER would be the Keystone server, which would then be used as the Keystone URL
+    Since we are using the chameleon portal, we want Horizon to use its preconfigured OPENSTACK_KEYSTONE_URL
+"""
+@sensitive_post_parameters()
+@csrf_exempt
+@never_cache
+def cc_websso(request):
+    """Logs a user in using a token from Keystone's POST."""
+    request.META['HTTP_REFERER'] = settings.OPENSTACK_KEYSTONE_URL
```

Comment:
Why do we set the referer here? I wouldn't think Keystone would care about this. It seems like it could be simplified to just

```python
    keystone_url = getattr(settings, 'OPENSTACK_KEYSTONE_URL')
    auth_url = utils.clean_up_auth_url(keystone_url)
    # Do stuff with auth_url
```

Reply:
I'm basically setting it just to maintain consistency with the original code (which expects Keystone url in referrer), I wasn't sure if the Keystone url was pulled from the request anywhere else, but it's really not likely and could probably be removed and simplified like you suggest.